coverage/docs/design.md

```diff
@@ -0,0 +1,156 @@
+produce a list of files that we care about in the line-by-line coverage report. produce line by 
+line coverage html and add link to covbot report. Note that covbot is the robot github account 
+used to report code coverage change results. See Covbot section for more details.
+5. If any file in this commit has a coverage change, let covbot post presubmit coverage on github, under that conversation of the PR. 
+  - The covbot comment should have the following information on each file with a coverage change
+    - file name
+    - old coverage (coverage before any change in the PR)
+    - new coverage (coverage after applied all changes in the PR)
+    - change the coverage
+  - After each new posting, any previous posting by covbot should be removed
+
+## Locally running presubmit and post-submit workflows
+Both workflows may be triggered locally in command line, as long as all the required flags are 
+supplied correctly. In addition, the following env var needs to be set:
+- JOB_TYPE (one of 'presubmit', 'postsubmit', 'local-presubmit')
+
+use 'local-presubmit' will run the presubmit workflow without posting result on github PR
+  
+## Covbot
+As mentioned in the presubmit workflow section, covbot is the short name for the robot github 
+account used to report code coverage change results. It can be created as a regular github 
+account - it does not need to be named covbot as that name is already taken on Github. It only need a 
+comment access to the repo it need to be run on. If the repo is private, it also need read access. 
+  
+After the robot account is created, download the github token and supply the path to the token 
+file to code coverage binary, as the value for parameter "github-token"
+
+
+# Usage with prow
+We pack the test coverage feature in a container, that is triggered to run by a CI/CD system such as prow, in response to Github events such as pulls and merges. The behavior varies in presubmit and postsubmit workflows, which is discussed in latter sections.
+
+Prow is a CI/CD system that is tested working well with this Code Coverage tool.   
+- Prow can be used as the system to handle Github events mentioned in the two workflows. 
+- Prow, in both workflows, supplies the flags & secrets for the binary, clones the repository and uploads logs & artifacts to GCS bucket.
+
+  - The pre-submit prow job is triggered by any new commit to a PR. At the end of the binary run, it can return a pass or fail status context to Github. Tide can use that status to block PR with low coverage.
+
+  - The post-submit prow job is triggered by merge events to the base branch.
+
+## Prow Job Specification
+Here is an example of a pre-submit prow job spec that runs the coverage tool in a container (the container build file can be found [here](https://github.com/kubernetes/test-infra/blob/a1e910ae6811a1821ad98fa28e6fad03972a8c20/coverage/Makefile)). The args section includes all the arguments for the binary of the tool. 
```

Comment:
The container image & Makefile are not really part of the code coverage feature either. They are there because prow needs to use them. Does other CI/CD systems also demand the use of container instead of the compiled binary?

Reply:
I think most do, or allow you to run binaries in a VM. Containerizing binaries for use in CI/CD is pretty standard practice though and should work with most other CI/CD systems. I would consider this step a relatively general step that is probably needed (or at least helpful) for running the coverage job in CI.